types/index.d.ts

```diff
@@ -0,0 +1,496 @@
+  }
+
+  export interface RelationshipOptions {
+    inverse?: string | null;
+    polymorphic?: boolean;
+  }
+
+  /** The registry-aware type of Mirage's `belongsTo` function */
+  // prettier-ignore
+  export type BelongsTo<Registry> =
+      & (<K extends keyof Registry>(key?: K, options?: RelationshipOptions) => () => Registry[K] | undefined)
+      & (<K extends keyof Registry>(options: RelationshipOptions & { polymorphic: true }) => () => Registry[K] | undefined);
+
+  /** Declares a one-to-one relationship to another Mirage model type. */
+  export function belongsTo<T = ModelInstance>(
+    key?: string,
+    options?: RelationshipOptions
+  ): () => T | undefined;
+  export function belongsTo<T = ModelInstance>(
+    options?: RelationshipOptions
+  ): () => T | undefined;
+
+  /** The registry-aware type of Mirage's `hasMany` function */
+  // prettier-ignore
+  export type HasMany<Registry> =
+      & (<K extends keyof Registry>(key?: K, options?: RelationshipOptions) => () => Collection<Registry[K]>)
+      & (<K extends keyof Registry>(options: RelationshipOptions & { polymorphic: true }) => () => Collection<Registry[K]>);
+
+  /** Declares a one-to-many relationship to another Mirage model type. */
+  export function hasMany<T = ModelInstance>(
+    key?: string,
+    options?: RelationshipOptions
+  ): () => Collection<T>;
+  export function hasMany<T = ModelInstance>(
+    options?: RelationshipOptions
+  ): () => Collection<T>;
+}
+
+declare module "miragejs/-types" {
+  // Captures the result of a `Model.extend()` call
+  const ModelData: unique symbol;
+  interface ModelDefinition<Data = {}> {
+    [ModelData]: Data;
+    extend<NewData>(data: NewData): ModelDefinition<Assign<Data, NewData>>;
+  }
+
+  // Captures the result of a `Factory.extend()` call
+  const FactoryData: unique symbol;
+  interface FactoryDefinition<Data = {}> {
+    [FactoryData]: Data;
+    extend<NewData>(data: NewData): FactoryDefinition<Assign<Data, NewData>>;
+  }
+
+  // The type-level equivalent of `Object.assign`
+  type Assign<T, U> = U & Omit<T, keyof U>;
+
+  // Extract relationship values from a model definition
+  type FlattenRelationships<T> = {
+    [K in keyof T]: T[K] extends () => infer Value ? Value : T[K];
+  };
+
+  // Extract factory method return values from a factory definition
+  type FlattenFactoryMethods<T> = {
+    [K in keyof T]: T[K] extends (n: number) => infer V ? V : T[K];
+  };
+
+  // Extracts model definition info for the given key, if a corresponding model is defined
+  type ExtractModelData<Models, K> = K extends keyof Models
+    ? Models[K] extends ModelDefinition<infer Data>
+      ? FlattenRelationships<Data>
+      : {}
+    : {};
+
+  // Extracts factory definition info for the given key, if a corresponding factory is defined
+  type ExtractFactoryData<Factories, K> = K extends keyof Factories
+    ? Factories[K] extends FactoryDefinition<infer Data>
+      ? FlattenFactoryMethods<Data>
+      : {}
+    : {};
```

Comment:
@dfreeman what do you think?

Reply:
The conditional check has to happen _somewhere_, because there might be entities that only have a model but not a factory defined, or vice versa. The overall type info for a Mirage record is determined by [the combination of the two](https://github.com/miragejs/miragejs/pull/262/files#diff-88bec6beae84369c0376b56b3bb88fe1R176-R178).

Reply:
@dfreeman should we change something here?

Reply:
Per my comment above, I don't think so, but @chriskrycho may have another suggestion.

Reply:
Nope, this is reasonable, I just wanted to make sure! 😅